src/kubernetes/reflector.rs

```diff
@@ -24,26 +24,53 @@ pub async fn custom_reflector<K, W>(
     loop {
         tokio::select! {
-            Some(Ok(event)) = stream.next() => {
-                match event {
-                    watcher::Event::Applied(_) => {
-                        // Immediately apply event
-                        store.apply_watcher_event(&event);
-                    }
-                    watcher::Event::Deleted(_) => {
-                        // Delay reconciling any `Deleted` events
-                        delay_queue.insert(event.to_owned(), delay_deletion);
-                    }
-                    watcher::Event::Restarted(_) => {
-                        // Clear all delayed events when the cache is refreshed
-                        delay_queue.clear();
-                        store.apply_watcher_event(&event);
-                    }
+            result = stream.next() => {
+                match result {
+                    Some(Ok(event)) => {
+                        match event {
+                            // Immediately reoncile `Applied` event
+                            watcher::Event::Applied(_) => {
+                                trace!("Processing Applied event.");
+                                store.apply_watcher_event(&event);
+                            }
+                            // Delay reconciling any `Deleted` events
+                            watcher::Event::Deleted(_) => {
+                                trace!("Queuing Deleted event.");
+                                delay_queue.insert(event.to_owned(), delay_deletion);
+                            }
+                            // Clear all delayed events on `Restarted` events
+                            watcher::Event::Restarted(_) => {
+                                trace!("Processing Restarted event.");
+                                delay_queue.clear();
+                                store.apply_watcher_event(&event);
+                            }
+                        }
+                    },
+                    Some(Err(error)) => {
+                        warn!(message = "Watcher stream got an error.", error = ?error);
+                    },
+                    None => {
+                        info!("Watcher stream has terminated.");
+                        break;
+                    },
                 }
             }
-            // When the Deleted event expires from the queue pass it to the store
-            Some(Ok(event)) = delay_queue.next() => {
-                store.apply_watcher_event(&event.into_inner())
+            result = delay_queue.next(), if !delay_queue.is_empty() => {
+                match result {
+                    Some(Ok(event)) => {
+                        trace!("Processing Deleted event.");
+                        store.apply_watcher_event(&event.into_inner());
+                    },
+                    Some(Err(error)) => {
+                        warn!(message = "DelayQueue stream got an error.", error = ?error);
+                    },
```

Comment:
I'm still really wondering about when this would happen 🤔  I can't figure out why the stream implementation would be yielding `Result`s.

Reply:
Same! But figure worst case we can log it and keep going 😅

Reply:
Mmm, I see this is actually fixed in newer versions of `tokio-util`: https://github.com/tokio-rs/tokio/commit/cf38ba627abb0d44ae7e7dc0659ddd51866e9efa

It also looks like @tobz approved it 😂 https://github.com/tokio-rs/tokio/pull/4241

It seems like it previously never returned an `Err`. I'd be ok with making this `unrechable()` too, with a comment.